opgaver and forløbsskabeloner draft

nk-randers opened the pull request:
**Models/DB**
* Added SQLAlchemy ORM model classes for Opgaver, Forløbsskabelon and Forløb to map to MSSQL DB
* Opgaver model has a many to one relationship with Forløbsskabelon and Forløb

**Utils/database and db_connection**
* Refactored DatabaseClient
* added utils for creation of DatabaseClient

**Endpoints for Opgaver**
* Get alle opgaver med en specifik forløbsID
* Get alle opgaver med en specifik forløbsskabelonID
* Delete en specfik opgave
* Put en specifik opgave
* Post en opgave med forløbskabelonID
* Post en opgave med forløbID

**Endpoints for Forløbsskabelon**
* GET alle Forløbsskabeloner
* POST en Forløbsskabelon
* PUT til at ændre navnet på skabelonen


**Endpoint for Forløb**
Only added this POST endpoint for forløb only for testing purposes to test the "Post en opgave med forløbID" endpoint from opgaver

nk-randers committed:
update config to use mssql db

nk-randers committed:
Update DatabaseClient

using sqlalchemy session

nk-randers committed:
fix lint

nk-randers committed:
added opgaver as model for table

using sqlalchemy as ORM to map data to Database

nk-randers committed:
added endpoints for opgaver

added GET and POST for /opgaver

nk-randers committed:
Added endpoints for opgaver

added PUT and DELETE

nk-randers committed:
update endpoint

get specific task endpoint added

nk-randers committed:
lint fix

nk-randers committed:
added base module

Add base module for SQLAlchemy models

nk-randers committed:
Updated Opgaver model

nk-randers committed:
Added model for Forløbsskabelon

nk-randers committed:
added controller for Opgaver

nk-randers committed:
added controller for Forløbsskabelon

nk-randers committed:
Refactor api_endpoints to use controller functions

nk-randers committed:
added put endpoint for opgave

nk-randers committed:
updated db table setup

nk-randers committed:
fix imports

nk-randers committed:
delete endpoint for opgaver

delete a specific opgave

nk-randers committed:
added utils function for db

nk-randers committed:
lint fix

nk-randers committed:
added endpoints for forløbskabelon

added get and post endpoint for forløbskabelon

nk-randers committed:
lint fix

nk-randers committed:
add put endpoint to change name

nk-randers committed:
fix lint

nk-randers committed:
added endpoints for Opgaver

endpoint for getting all opgaver with a spceific forløbsid and post endpoint

nk-randers committed:
added post endpoint for forløb

st-randers commented on python/src/models.py:
```diff
@@ -0,0 +1,41 @@
+    OpgaverID = Column(Integer, primary_key=True, autoincrement=True)
+    title = Column(String, nullable=False)
+    beskrivelse = Column(String, nullable=False)
+    resourcer = Column(String, nullable=False)
```
Hvordan opbevarer vi ressourcer? Vi bør nok lave en ny klasse/tabel til ressourcer med (opgaveID, navn, url) som properties.

st-randers commented on python/src/models.py:
```diff
@@ -0,0 +1,41 @@
+    opgaver = relationship('Opgaver', back_populates='forløb')
+
+
+class Opgaver(Base):
```
Bør nok hedde "opgave" i stedet for "opgaver"

st-randers commented on python/src/controllers/opgaver_controller.py:
```diff
@@ -0,0 +1,184 @@
+        session.close()
+
+
+def create_opgaver_with_forloeb_id():
```
Jeg er ikke sikker på jeg forstår formålet med denne funktion. Som jeg ser det, skal vi bruge én funktion til at oprette et nyt blank forløb, én funtion til at oprette et nyt forløb med skabelonID (se ovenstående kommentar), én funktion til at oprette en ny blank opgave, og en funktion til at oprette en ny opgave med et opgaveskabelonId.

st-randers commented on python/src/controllers/opgaver_controller.py:
```diff
@@ -0,0 +1,184 @@
+db_client = get_db_client()
+
+
+def create_opgaver_with_forloebsskabelon_id():
```
Kopierer den alle opgaver fra et forløbsskabelon til et nyt forløb? Jeg tænker flowet er at man enten opretter et blank forløb, eller opretter et forløb med en forløbsskabelon. I dette tilfælde skal funktionen oprette et nyt forløb, og tilføje alle opgaver fra skabelonen (dvs. alle opgaver med dennes forløbsskabelonId).

st-randers commented on python/src/controllers/opgaver_controller.py:
```diff
@@ -0,0 +1,184 @@
+        session.close()
+
+
+def get_all_opgaver():
```
Ændr til get_all_opgaveskabeloner. Vi har ikke brug for at trække alle opgaver på én gang, kun alle opgaver med specifikt forloebsID

st-randers reviewed (changes requested):
Har smidt et par kommentarer til koden. 

Jeg tænker at der mangler disse funktioner:
- create_forloeb (med forloebsskabelonID) -> kopierer alle opgaver fra forløbsskabelonen til det nye forløb
- create_opgave (blank opgave) -> tilknyttes forløb ELLER forløbsskabelon
- create_opgave (med opgaveskabelonID) -> tilknyttes forløb ELLER forløbsskabelon
- get_all_opgaveskabeloner()

Derudover synes jeg alle tilfælde af "opgaver" skal ændres til "opgave" når der er tale om én opgave. Hvis funktionen henter flere opgaver, er det fint den hedder "opgaver".

Til sidst, skal vi nok have tilføjet en "ressource"-klasse, som skal indeholde ressourcer tilknyttet bestemte opgaver. Se inline kommentar herom. Hertil bygges funktioner:
- create_ressource() -> tilknyttes opgave eller opgaveskabelon
- delete_ressource()
- edit_ressource()
- get_ressources_by_opgaveid()
- get_ressources_by_opgaveskabelonid()
**ALTERNATIVT indbygges ressourcer i funktion som henter opgave-data, således det ikke kræver et ekstra GET-funktions/API-kald.**

nk-randers committed:
endpoint for creating forloeb

Copy all opagver from the forløbsskabelon to the new forloeb

nk-randers committed:
updated create_opgave endpoint

create_opgave endpoint is now associated with forloeb or forloeb template

nk-randers committed:
Added Opgaveskabeloner and Ressource

Added models Opgaveskabeloner and Ressource

nk-randers committed:
added endpoints for Opgaveskabelon

added get, post, put and delete for Opgaveskabelon

nk-randers committed:
Update name to Opgave

Update model name from Opgaver to Opgave

nk-randers committed:
Update opgave_controller

Added endpoint for creating opgave with a opgave template associated with either a forloeb or forloeb template. also copies resources associated with the opgaveskabelon

nk-randers committed:
added endpoints for ressource

added 5 endpoints for ressource

nk-randers committed:
lint fix

nk-randers committed:
lint fix

nk-randers committed:
Update api_endpoints.py

nk-randers committed:
update controller name

nk-randers committed:
Update models.py

Columns in the Forløb table are now required and cannot be null.

st-randers reviewed (approved):
Ser rigtig fornuftigt ud! Kan ikke umiddelbart se at der mangler noget, så super godt.
